File: calc_parser.py
"""

"""



from enum import Enum,auto

class func:
    def __init__(self,funccode:str):
        self.code:str=funccode
        funcname, funcargs=self.split()
        self.funcname=self.check_funcname(funcname)
        if " " in self.funcname:
            raise BaseException("関数名に半角スペースを入れないでください")
        self.funcargs=self.split_args(funcargs)
    @classmethod
    def new(cls,code:str):return func(code)
    def check_funcname(self,text:str)->str:
        #funcname内の文字列がすべて有効かどうかを調べる関数
        if text[0]==" ":#半角スペースが文字列の前に入っていた場合
            return self.check_funcname(text[1:])
        if text[-1]==" ":
            return self.check_funcname(text[:-1])
        return text

    def split(self)->tuple[str,str]:
        index = self.code.find("(")
        if index<0:
            raise BaseException("引数部分の文法が不正だと思われます。おそらく")
        else:
            funcname:str = self.code[:index]
            funcargs:str = self.code[index:]
        return funcname,funcargs
    def split_args(self,funcargs):
        #引数を抽出する
        depth:int=0
        args:list[str]=[]
        arg:str=""
        for i in funcargs:
            match i:
                case "(":
                    if depth>0:
                        arg+=i
                    depth+=1
                case ")":
                    depth-=1
                    if depth>0:
                        arg+=i
                case _:
                    if depth>1:
                        arg+=i
                    else:
                        if i==",":
                            args.append(arg)
                            arg=""
                        else:
                            arg+=i
        if arg:
            args.append(arg)
        return args
    @property
    def data(self):
        funcname,funcargs = self.split()
        funcname = self.check_funcname(funcname)
        return [funcname]+self.split_args(funcargs)
    def __repr__(self) -> str:
        return f"name:'{self.funcname}'\nargs:{self.funcargs}"

class brackets:
    def __init__(self,bracketscode:str):
        self.code:str=bracketscode
    @classmethod
    def new(cls,code):return brackets(code)

    def __inner_content(self,code:str):
        depth:int=0
        rlist:list[str]=[]
        for i in code:
            match i:
                case "(":
                    if depth>0:rlist.append(i)
                    depth+=1
                case ")":
                    depth-=1
                    if depth>0:rlist.append(i)
                case _:
                    if depth>0:rlist.append(i)
        rtext = "".join(rlist)
        return self.__inner_content(rtext) if elem.new(rtext).elemtype is Elem_type.BRACKETS else rtext
    @property
    def inner(self)->str:
        return self.__inner_content(self.code)

class value:
    def __init__(self,code:str):
        self.code:str=self.check_valuename(code)
        if " " in self.code:
            raise BaseException("半角スペースが入っています")
    @classmethod
    def new(cls,code:str):return value(code)
    def check_valuename(self,valuename:str)->str:
        if valuename[0]==" ":
            return self.check_valuename(valuename[1:])
        if valuename[-1]==" ":
            return self.check_valuename(valuename[:-1])
        return valuename
    @property
    def inner(self):
        return self.code

class Elem_type(Enum):
    #type of elem
    FUNCTION  = auto()
    OPERATION = auto()
    NUMBER    = auto()
    BRACKETS  = auto()
    OTHER     = auto()
    VALUE     = auto()
    FORMULA   = auto()

class elem:
    #式のelement
    def __init__(self,code):
        """

        """
        self.availablechars:list[str]=[
                chr(i) for i in range(65,65+26)#A~Z
        ]+[
                chr(i) for i in range(97,97+26)#a~z
        ]+[
                str(i) for i in range(10)      #0~9
        ]
        self.code=code
    @classmethod
    def new(cls,code):return elem(code)

    #bool
    def __is_FUNCTION(self,code:str)->bool:#工事中
        index = code.find("(")
        if index<0:
            return False
        else:
            funcname:str = code[:index]
            funcargs:str = code[index:]
            #本当はここに関数名が正しく設定されているかを確認する関数を作成したい
            if self.__is_BRACKETS(funcargs):
                return True
            else:
                return False
    def __is_BRACKETS(self,code:str)->bool:
        group:list[bool]=[]
        depth:int=0
        for i in code:
            match i:
                case "(":
                    depth+=1
                    group.append(depth>0)
                case ")":
                    group.append(depth>0)
                    depth-=1
                case " ":
                    pass
                case _:
                    group.append(depth>0)
        return all(group)
    def __is_NUMBER(self,code:str)->bool:
        #⚠️少数点の場合も含む
        #アンダースコアは許容しない(とりあえずは。。。)
        group:list[str]=[]
        nums:list[str]=list(map(str,range(10)))#1~10の数字のリスト
        for i in code:
            if i in nums:
                group.append(i)
            elif i==".":
                if group:
                    group.append(i)
                else:
                    raise BaseException("小数点が先頭についてしまっていますよ")
            else:
                return False

        return True
    def __is_OPERATION(self,code:str)->bool:
        opelist:list[str] = ["+","-","*","/","%","^","@"]
        for i in opelist:
            if i==code:
                return True
        return False
    def __is_VALUE(self,code:str)->bool:
        for i in code:
            if i in self.availablechars:
                pass
            elif i==" ":
                pass
            else:
                return False
        return True
    def __is_FORMULA(self,code:str)->bool:
        depth:int=0
        for i in code:
            match i:
                case "(":
                    depth+=1
                case ")":
                    depth-=1
                case _:
                    if depth>0:
                        pass
                    else:
                        if i in ["+","-","*","/","%","^","@"]:
                            return True
        return False

    @property
    def elemtype(self)->Elem_type:
        if self.__is_OPERATION(self.code):
            return Elem_type.OPERATION
        elif self.__is_NUMBER(self.code):
            return Elem_type.NUMBER
        elif self.__is_BRACKETS(self.code):
            return Elem_type.BRACKETS
        elif self.__is_FORMULA(self.code):
            return Elem_type.FORMULA
        elif self.__is_FUNCTION(self.code):
            return Elem_type.FUNCTION
        elif self.__is_VALUE(self.code):
            return Elem_type.VALUE
        else:
            return Elem_type.OTHER

class parser:
    def __init__(self,code:str):
        self.code:str=code
        self.availablechars:list[str]=[
                chr(i) for i in range(65,65+26)#A~Z
        ]+[
                chr(i) for i in range(97,97+26)#a~z
        ]+[
                str(i) for i in range(10)      #0~9
        ]
    def grouping_number(self,vec:str)->list[str]:
        #numberをまとめる
        rlist:list[str]=list()
        group:list[str]=list()
        flag:bool=False
        for i in vec:
            if i in self.availablechars:
                group.append(i)
                flag=True
            else:
                if flag:
                    rlist.append("".join(group))
                    group=[]
                    flag=False
                rlist.append(i)
        if flag:
            rlist.append("".join(group))
        return rlist
    def grouping_brackets(self,vec:list[str])->list[str]:
        #functionをまとめる
        rlist:list[str]=list()
        group:list[str]=list()
        depth:int=0
        flag:bool=False
        for i in vec:
            match i:
                case "(":
                    if depth>0:
                        group.append(i)
                    elif depth==0:
                        group.append(i)
                        flag=True
                    else:
                        raise BaseException("括弧を閉じ忘れている可能性があります")
                    depth+=1
                case ")":
                    depth-=1
                    if depth>0:
                        group.append(i)
                    elif depth==0:
                        group.append(i)
                        rlist.append("".join(group))
                        group=[]
                        flag=False
                    else:
                        raise BaseException("括弧を開き過ぎている可能性があります")
                case _:
                    if flag:
                        group.append(i)
                    else:
                        rlist.append(i)

        return rlist
    def split_symbol(self,vec:list[str])->list[str]:
        rlist:list[str]=list()
        group:list[str]=list()
        for i in vec:
            match i:
                case "+"|"-"|"*"|"/"|"%"|"@"|"^":
                    if group:#groupがからでなければ
                        rlist.append("".join(group))
                        group=list()
                    rlist.append(i)
                case _:
                    group.append(i)
        if group:
            rlist.append("".join(group))
        return rlist
    def code2vec(self)->list[str]:
        vec=self.grouping_number(self.code)
        vec=self.grouping_brackets(vec)
        vec=self.split_symbol(vec)
        return vec

    def is_Formula(self,code:str)->bool:
        return elem.new(code).elemtype is Elem_type.FORMULA
    def resolve_operation(self,code:str):
        #逆ポーランド記法の返り値
        #引数が二つであると決定した
        par = parser(code)
        vec = par.code2vec()
        index = self.priority(vec)
        if index is None:
            return code
        E1:str= "".join(vec[:index])
        ope = vec[index]
        E2:str= "".join(vec[index+1:])
        #return [ ope, *self.resolve_util(E1), *self.resolve_util(E2)]
        return formula_tree(
                ope,
                Elem_type.OPERATION,
                [*self.resolve_util(E1),*self.resolve_util(E2)]
            )
    def resolve_util(self,E):
        return self.resolve_operation(E)\
            if self.is_Formula(E) else self.resolve_operation(brackets.new(E).inner)\
            if elem.new(E).elemtype is Elem_type.BRACKETS else self.resolve_function(E)\
            if elem.new(E).elemtype is Elem_type.FUNCTION else value.new(E).inner\
            if elem.new(E).elemtype is Elem_type.VALUE else E,#カンマとるなよ絶対に！
    def resolve_function(self,code:str):
        funcdata = func.new(code).data
        funcname=funcdata[0]
        #return [funcname]+[self.resolve_util(i)[0] for i in funcdata[1:]]
        return formula_tree(
            funcname,
            Elem_type.FUNCTION,
            [self.resolve_util(i)[0] for i in funcdata[1:]]
        )

    def resolve(self):
        return self.resolve_util(self.code)
    def priority(self,vec:list[str])->int|None:
        #配列内にある最も優先順位が低い演算子を探します
        #返り値はindexです
        """
        実装されている計算規則
        rankの数字が低いものほど計算順位も低い
        同一のランクが並列に続く場合、式の右に行く程計算順位が低い
        """
        rankinglist:dict={
            #演算子優先順位
            "+":1,"-":1,
            "*":2,"/":2,"%":2,"@":2,
            "^":3
        }
        rank:int=3
        index=None
        for i,j in enumerate(vec):
            if j in rankinglist.keys():
                if rank>rankinglist[j]:
                    rank=rankinglist[j]
                    index=i
                elif rank==rankinglist[j]:
                    #右に計算順位が同じ演算子を見つけた場合
                    #右にあるものの方が計算順位が低い
                    rank=rankinglist[j]
                    index=i
                else:
                    pass
            else:
                pass
        #最も計算順位が1番低いindexを返す
        return index

class formula_tree:
    def __init__(self,name:str,type_:Elem_type,args:list):
        self.name:str=self.ope2func(name)
        self.args:list=args
        self.selftype:Elem_type=type_
    def ope2func(self,name:str)->str:
        """
        演算子を関数のように見る
        """
        match name:
            case "+":
                return "+"
            case "-":
                return "-"
            case "*":
                return "*"
            case "/":
                return "/"
            case "%":
                return "%"
            case "@":
                return "@"
            case _:
                return name
    def __getitem__(self,key):
        return self.args[key]
    def __repr__(self):
        return f"{self.name}{self.args}"


if __name__=="__main__":
    """texts=[
    " 10 + ( x + log10(2) * sin(x) ) * log10(x)",
    "(-sin(x)*3)+(-2*cos(x))",
    "pi",
    "sin(x)",
    "(1)+2",
    "3.14",
    "-8+6"
    ]
    """
    texts = [
        "main(a,b,c)"
    ]
    for i in texts:
        par=parser(i)
        print(par.resolve()[0])
    el = elem("-sin(x)")
    print(el.elemtype)
    #print(
    #    elem.new(" sin(x) ").elemtype
    #)


"""
fn f(){
    return g()
}
fn g():
    return f()
"""

File: calc_parser.js
/*
#calc_parser.js
## written by Tom0427
## for Google App Script





*/

const Elem_type = {
    //列挙型(のつもり)
    //
    FUNCTION : 0,
    OPERATION: 1,
    NUMBER   : 2,
    BRACKETS : 3,
    OTHER    : 4,
    VALUE    : 5,
    FORMULA  : 6
}

class func{
    constructor(funccode){
        this.code = funccode;
        let funcelem = this.split();
        this.funcname = this.check_funcname(funcelem[0]);
        if (this.funcname.includes(" ")){
            throw new Error("関数名にスペースを入れないでください");
        }
        this.funcargs = this.split_args(funcelem[1]);
    }
    check_funcname(text){
        if (text[0]==" "){
            return this.check_funcname(text.slice(1));
        }
        if (text[text.length-1]==" "){
            return this.check_funcname(text.slice(0,-1));
        }
        return text;
    }

    split(){
        let funcname;
        let funcargs;
        let index = [...this.code].findIndex(element=>element==="(");
        if (index<0){
            throw new Error("引数部分の文法が不正です");
        }
        else{
            funcname = this.code.slice(0,index);
            funcargs = this.code.slice(index);
        }
        return [funcname,funcargs];
    }
    split_args(funcargs){
        let depth = 0;
        let args = [];
        let arg = "";
        for (const i of funcargs){
            switch (i){
                case "(":
                    if (depth>0){
                        arg+=i
                    }
                    depth++;
                    break;
                case ")":
                    depth--;
                    if (depth>0){
                        arg+=i;
                    }
                    break;
                default:
                    if (depth>1){
                        arg+=i;
                    }else{
                        if (i==","){
                            args.push(arg);
                            arg="";
                        }else{
                            arg+=i;
                        }
                    }
            }
        }
        if (arg.length!=0){
            args.push(arg);
        }
        return args
    }
    data(){
        let funcelems = this.split()
        let funcname = funcelems[0];
        let funcargs = funcelems[1];
        funcname = this.check_funcname(funcname)
        return [funcname,...this.split_args(funcargs)]
    }
}

class brackets{
    constructor(bracketscode){
        this.code = bracketscode;
    }
    __inner_content(code){
        let depth = 0;
        let rlist = [];
        for (const i of code){
            switch(i){
                case "(":
                    if(depth>0){
                        rlist.push(i);
                    }
                    depth++;
                    break;
                case ")":
                    depth--;
                    if (depth>0){
                        rlist.push(i);
                    }
                    break;
                default:
                    if (depth>0){
                        rlist.push(i)
                    }
            }
        }
        let rtext = rlist.join("");
        let element=new elem(rtext);
        return (element.elemtype() ===Elem_type.BRACKETS?this.__inner_content(rtext):rtext);
    }
    inner(){
        return this.__inner_content(this.code);
    }
}

class value{
    constructor(code){
        this.code = this.check_funcname(code);
        if (this.code.includes(" ")){
            //throw new Error("変数内にスペースを入れてはいけません")
        }
    }
    check_funcname(text){
        if (text[0]==" "){
            return this.check_funcname(text.slice(1));
        }
        if (text[text.length-1]==" "){
            return this.check_funcname(text.slice(0,-1));
        }
        return text;
    }
    inner(){
        return this.code;
    }
}

class elem{
    constructor(code){
        this.availablechars = [
            ...[...Array(26)].map((_, i) => i).map(a => String.fromCodePoint(a + 65)),
            ...[...Array(26)].map((_, i) => i).map(a => String.fromCodePoint(a + 97)),
            ...[...Array(10)].map((_, i) => i).map(a => a.toString())
        ];
        this.code = code;
    }
    __is_FUNCTION(code){
        let index = [...code].findIndex((char)=>char=="(");
        if (index<0){
            return false;
        }else{
            let funcname = code.slice(0,index);
            let funcargs = code.slice(index);
            if(this.__is_BRACKETS(funcargs)){
                return true;
            }else{
                return false;
            }
        }
    }
    __is_BRACKETS(code){
        let group = []
        let depth = 0
        for (const i of code){
            switch (i){
                case "(":
                    depth++;
                    group.push(depth>0)
                    break;
                case ")":
                    group.push(depth>0)
                    depth--;
                    break;
                case " ":break;
                default:
                    group.push(depth>0)
            }
        }
        return group.every(a => a===true);
    }
    __is_NUMBER(code){
        let group = [];
        let nums = [...Array(10)].map((_,i)=>i).map(i=>i.toString());
        for (const i of code){
            if (nums.includes(i)){
                group.push(i)
            }else if (i=="."){
                if (group.length!=0){
                    group.push(i)
                }else{
                    throw new Error("小数点、先頭についてますよ！♡");
                }
            }else{
                return false;
            }
        }
        return true;
    }

    __is__OPERATION(code){
        let opelist = ["+", "-", "*", "/", "%", "^", "@"];
        for (const i of opelist){
            if (i===code){
                return true;
            }
        }
        return false;
    }
    __is_VALUE(code){
        for (const i of code){
            if (this.availablechars.includes(i)){
                //pass
            }else if (i==" "){
                //pass
            }else{
                return false;
            }
        }
        return true;
    }
    __is_FORMULA(code){
        let depth = 0
        let opelist = ["+", "-", "*", "/", "%", "^", "@"];
        for (const i of code){
            switch(i){
                case "(":
                    depth++;
                    break;
                case ")":
                    depth--;
                    break;
                default:
                    if (depth>0){
                        //pass
                    }else{
                        if (opelist.includes(i)){
                            return true;
                        }
                    }
            }
        }
        return false;
    }
    elemtype(){
        if (this.__is__OPERATION(this.code)){
            return Elem_type.OPERATION;
        }
        else if (this.__is_NUMBER(this.code)){
            return Elem_type.NUMBER;
        }
        else if (this.__is_BRACKETS(this.code)){
            return Elem_type.BRACKETS;
        }
        else if (this.__is_FORMULA(this.code)){
            return Elem_type.FORMULA;
        }
        else if (this.__is_FUNCTION(this.code)){
            return Elem_type.FUNCTION;
        }
        else if (this.__is_VALUE(this.code)){
            return Elem_type.VALUE;
        }else{
            return Elem_type.OTHER
        }
    }
}

class parser{
    
    constructor(code){
        this.code = code;
        this.availablechars = [
            ...[...Array(26)].map((_, i) => i).map(a => String.fromCodePoint(a + 65)),
            ...[...Array(26)].map((_, i) => i).map(a => String.fromCodePoint(a + 97)),
            ...[...Array(10)].map((_, i) => i).map(a => a.toString())
        ];
    }

    grouping_numbers(vec){
        let rlist = [];
        let group = [];
        let flag = false;
        for (const i of vec){
            if (this.availablechars.includes(i)){
                group.push(i);
                flag=true;
            }else{
                if (flag){
                    rlist.push(group.join(""));
                    group=[];
                    flag = false;
                }
                rlist.push(i);
            }
        }
        if (flag){
            rlist.push(group.join(""));
        }
        return rlist;
    }

    grouping_brackets(vec){
        let rlist = [];
        let group = [];
        let depth = 0;
        let flag = false;
        for (const i of vec){
            switch (i){
                case "(":
                    if (depth>0){
                        group.push(i);
                    }else if (depth==0){
                        group.push(i);
                        flag = true;
                    }else{
                        throw new Error('括弧を閉じ忘れている可能性があります');
                    }
                    depth++;
                    break;
                case ")":
                    depth--;
                    if (depth>0){
                        group.push(i);
                    }else if (depth==0){
                        group.push(i);
                        rlist.push(group.join(""));
                        group = [];
                        flag = false;
                    }else{
                        throw new Error('括弧を開き過ぎている可能性があります');
                    }
                    break;
                default:
                    if (flag){
                        group.push(i);
                    }else{
                        rlist.push(i);
                    }
            }
        }
        return rlist;
    }

    split_symbol(vec){
        let rlist = [];
        let group = [];
        for (const i of vec){
            switch (i){
                case "+":
                case "-":
                case "*":
                case "/":
                case "%":
                case "@":
                case "^":
                    if (group.length!=0){
                        rlist.push(group.join(""));
                        group = [];
                    }
                    rlist.push(i);
                    break;
                default:
                    group.push(i);
            }
        }
        if (group!=0){
            rlist.push(group.join(""))
        }
        return rlist;
    }

    code2vec(){
        let vec = this.grouping_numbers(this.code);
        vec = this.grouping_brackets(vec);
        vec = this.split_symbol(vec);
        return vec;
    }

    isFormula(code){
        let element = new elem(code);
        return element.elemtype() === Elem_type.FORMULA;
    }

    resolve_operation(code){
        const par = new parser(code);
        const vec = par.code2vec();
        const index = this.priority(vec);
        if (index===null){
            return code;
        }
        const e1=vec.slice(0,index).join("");
        const ope=vec[index];
        const e2=vec.slice(index+1).join("");
        return [
            ope,
            this.resolve_util(e1),
            this.resolve_util(e2)
        ]
    }
    resolve_util(E){
        /*
        python内包表記より
        */
        let element = new elem(E);
        let type = element.elemtype();
        if (this.isFormula(E)){
            return this.resolve_operation(E);
        }else if (type===Elem_type.BRACKETS){
            let bra = new brackets(E)
            return this.resolve_operation(bra.inner())
        }else if (type===Elem_type.FUNCTION){
            return this.resolve_function(E);
        }else if (type===Elem_type.VALUE){
            let val = new value(E);
            return val.inner();
        }else{
            return E;
        }
    }

    resolve_function(code){
        let fun = new func(code);
        let funcdata = fun.data();
        let funcname = funcdata[0];
        return [
            funcname,
            ...funcdata.slice(1).map(
                i=>this.resolve_util(i)
            )
        ];
    }

    resolve(){
        return this.resolve_util(this.code);
    }
    
    priority(vec){
        let rankinglist={
            "+":1,"-":1,
            "*":2,"/":2,"%":2,"@":2,
            "^":3
        }
        let rank = 3;
        let index = null;
        for (const i in vec){
            const j = vec[i];
            if (Object.keys(rankinglist).includes(j)){
                if (rank>rankinglist[j]){
                    rank = rankinglist[j];
                    index = i; 
                }else if (rank==rankinglist[j]){
                    rank = rankinglist[j]
                    index=i
                }else{
                    //pass
                }
            }else{
                //pass
            }
        }
        if (index==null){
            return null
        }
        else{
            return Number(index);
        }
    }
}


class calcVM{
    constructor(resolved_data){
        this.resolved_data = resolved_data;
        this.constants={
            "pi":Math.PI,
            "e":Math.E,
        };
    }
    run(){
        return this.step(this.resolved_data);
    }
    step(arr){
        if (!Array.isArray(arr)){
            return this.constants2value(arr);
        }
        let funcname = arr[0];
        let args = arr.slice(1);
        //引数の前処理
        args = args.map((i)=>this.constants2value(i));
        args = args.map((i)=>Number(i));
        return this.functions(funcname,args);
    }
    constants2value(text){
        //再帰
        if (Array.isArray(text)){
            return this.step(text);
        }
        let element = new elem(text);
        if (element.elemtype()==Elem_type.NUMBER){
            return text;
        }else{
            let rtext = this.constants[text];
            if (rtext===undefined){
                throw new Error(`${text}は定数として定義されていません`);
            }
            return rtext;
        }
    }
    functions(funcname,args){
        switch (funcname){
            case "+":
                return args[0]+args[1];
            case "-":
                if (args==''){
                    return -1*args[1];
                }else{
                    return args[0]-args[1];
                }
            case "*":
                return args[0]*args[1];
            case "/":
                return args[0]/args[1];
            case "%":
                return args[0]%args[1];
            case "^":
                return args[0]**args[1];
            case "sqrt":
                return Math.sqrt(args[0]);
            case "sin":
                return Math.sin(args[0]);
            case "cos":
                return Math.cos(args[0]);
            case "tan":
                return Math.tan(args[0]);
            case "log":
                return Math.log(args[0]);
            case "log10":
                return Math.log10(args[0]);
            case "pow":
                return Math.pow(args[0],args[1]);
            case "floor":
                return Math.floor(args[0]);
            case "ceil":
                return Math.ceil(args[0]);
            case "random":
                return Math.random();
            case "randint":
                let min = Math.ceil(args[0]);
                let max = Math.floor(args[1]);
                return Math.floor(Math.random() * (max - min) + min); //The maximum is exclusive and the minimum is inclusive
            default:
                throw new Error(`指定された関数は${funcname}見つかりませんでした`)
        }
    }

}

function main(){
    let texts = [
        //"he llo(10,10)",
        "f((0,1"
    ]

    //for (const i of texts){
    //    let par = new parser(i);
    //    //console.log(JSON.stringify(par.resolve()));
    //    console.log(par.resolve());
    //}
    //FUNCTION : 0,
    //OPERATION: 1,
    //NUMBER   : 2,
    //BRACKETS : 3,
    //OTHER    : 4,
    //VALUE    : 5,
    //FORMULA  : 6
    for (const i of texts){
        let par = new parser(i);
        console.log(JSON.stringify(par.resolve()));
        try{
            let cvm = new calcVM(par.resolve());
            console.log(cvm.run());
        }catch (e){
            console.log(e)
        }
    }
    //let cvm = new calcVM(par.resolve());
    //console.log(cvm.run());
}

main()